### ISSUE TITLE
[BUG] Expansion error still occurs in 1.4.2

### ISSUE THREAD

**Issue by pwurbs** [longhorn/longhorn#6078]
## Describe the bug (🐛 if you encounter this issue)

The issue #5513 was documented as resolved in the release notes of 1.4.2.
Yesterday I upgraded from 1.4.1 to 1.4.2. 
This morning, after a node restart, the error occured again. 

```
Expansion Error: BUG: The expected size 2147483648 of engine pvc-97135cac-3890-42b1-bc32-80a0edec1b2e-e-6151baeb should not be smaller than the current size 5368709120
Note: You can cancel the expansion to avoid volume crash
```

Stopping the expansion throws an error, that the expansion has not started yet. 
Due to the node restart, we only have 1 replica instead of 3, because rebuild can't start. And we are not able to recover from this state except restoring from backup. 
Because there still seems to be a probability that this issue can occur at any time and there is no good way to recover, we have to stop the rollout for version 1.4.2

## To Reproduce

It's not really reproduceable. It happens from time to time after a node restart. 

## Expected behavior

The issue should not occur. It would be at least helpful to know how to recover from this state. 

## Environment

 - Longhorn version: 1.4.2
 - Installation method (e.g. Rancher Catalog App/Helm/Kubectl): Rancher  Catalog App
 - Kubernetes distro (e.g. RKE/K3s/EKS/OpenShift) and version: RKE1 / K8S v1.24.13
   - Number of management node in the cluster: 3
   - Number of worker node in the cluster: 4
 - Node config
   - OS type and version: CentOS 7
   - CPU per node: 16 vCPU
   - Memory per node: 64GB
   - Disk type(e.g. SSD/NVMe): SSD
   - Network bandwidth between the nodes: 2 - 6 Gbit/s
 - Underlying Infrastructure (e.g. on AWS/GCE, EKS/GKE, VMWare/KVM, Baremetal): Cloud/KVM
 - Number of Longhorn volumes in the cluster: 110

**Comment by ejweber** [comment]
I can think of one realistic possibility (besides the obvious possibility that there is some path to inappropriate expansion we haven't covered yet):

The "damage" had already been done previously by v1.4.1. Replicas could be inappropriately expanded, but still running fine until a node reboot. If this is the case, we may need to develop a way to proactively prune previously expanded replicas even with a working fix in v1.4.2.

If there are snapshots, we may be able to confirm this. Can you exec into the `instance-manager-r` pod or ssh into the node that has the badly sized replica, cd into the replica directory (either `/host/var/lib/longhorn/replicas/<name>` or `/var/lib/longhorn/replicas/<name>` depending on how you got there) and execute `ls -la` and `grep "" *.meta`?

**Comment by pwurbs** [comment]
> @pwurbs Could you provide the support bundle?
> 
> cc @ejweber

I would like to do, but there is an error in the support bundle daemon set:
`Error creating: pods "supportbundle-agent-support-bundle-2023-06-08t06-41-39z-" is forbidden: PodSecurityPolicy: unable to admit pod: [spec.volumes[0]: Invalid value: "hostPath": hostPath volumes are not allowed to be used spec.containers[0].securityContext.capabilities.add: Invalid value: "SYSLOG": capability may not be added spec.containers[0].securityContext.capabilities.add: Invalid value: "SYSLOG": capability may not be added] `

**Comment by pwurbs** [comment]
@ejweber 
Below the results of the command for one of the affected volumes.

ls -la
```
total 6880352
drwx------.   2 root root        4096 Jun  8 04:42 .
drwxr-xr-x. 109 root root       20480 Jun  8 07:44 ..
-rw-------.   1 root root        4096 Jun  8 07:50 revision.counter
-rw-r--r--.   1 root root 42949672960 Jun  8 07:50 volume-head-007.img
-rw-r--r--.   1 root root         187 Jun  8 04:42 volume-head-007.img.meta
-rw-r--r--.   1 root root 42949672960 Jun  8 04:39 volume-snap-dailybac-d0c4f62a-8f7a-4522-854e-c754e1dadeb9.img
-rw-r--r--.   1 root root         212 Jun  8 04:40 volume-snap-dailybac-d0c4f62a-8f7a-4522-854e-c754e1dadeb9.img.meta
-rw-r--r--.   1 root root 42949672960 Jun  7 12:03 volume-snap-snapshot-0236df7a-8b33-4569-8014-e33d735a4e01.img
-rw-r--r--.   1 root root         172 Jun  7 16:04 volume-snap-snapshot-0236df7a-8b33-4569-8014-e33d735a4e01.img.checksum
-rw-r--r--.   1 root root         210 Jun  7 12:03 volume-snap-snapshot-0236df7a-8b33-4569-8014-e33d735a4e01.img.meta
-rw-r--r--.   1 root root 42949672960 Jun  8 04:42 volume-snap-snapshot-23cbf46b-e1f8-41c7-8d21-edbdacdc38a0.img
-rw-r--r--.   1 root root         210 Jun  8 04:42 volume-snap-snapshot-23cbf46b-e1f8-41c7-8d21-edbdacdc38a0.img.meta
-rw-r--r--.   1 root root 42949672960 Jun  6 12:11 volume-snap-snapshot-488c080c-0b4f-442f-aeec-667cd36f58cb.img
-rw-r--r--.   1 root root         172 Jun  6 16:50 volume-snap-snapshot-488c080c-0b4f-442f-aeec-667cd36f58cb.img.checksum
-rw-r--r--.   1 root root         252 Jun  6 12:11 volume-snap-snapshot-488c080c-0b4f-442f-aeec-667cd36f58cb.img.meta
-rw-r--r--.   1 root root 42949672960 Jun  7 15:08 volume-snap-snapshot-60621c68-3dc8-445d-bc08-f0f3c5587416.img
-rw-r--r--.   1 root root         172 Jun  7 16:02 volume-snap-snapshot-60621c68-3dc8-445d-bc08-f0f3c5587416.img.checksum
-rw-r--r--.   1 root root         210 Jun  7 15:08 volume-snap-snapshot-60621c68-3dc8-445d-bc08-f0f3c5587416.img.meta
-rw-r--r--.   1 root root 10737418240 Jun  6 12:11 volume-snap-snapshot-65bfafe1-9581-496a-81bf-78a3151c658d.img
-rw-r--r--.   1 root root         172 Jun  6 16:52 volume-snap-snapshot-65bfafe1-9581-496a-81bf-78a3151c658d.img.checksum
-rw-r--r--.   1 root root         252 Jun  6 12:11 volume-snap-snapshot-65bfafe1-9581-496a-81bf-78a3151c658d.img.meta
-rw-r--r--.   1 root root 42949672960 Jun  8 04:40 volume-snap-snapshot-71db93c1-d06f-4689-9365-5892a4bfc642.img
-rw-r--r--.   1 root root         252 Jun  8 04:40 volume-snap-snapshot-71db93c1-d06f-4689-9365-5892a4bfc642.img.meta
-rw-r--r--.   1 root root 10737418240 Jun  8 04:42 volume-snap-snapshot-ab1a619f-196d-4f58-9a35-2c705a05cacb.img
-rw-r--r--.   1 root root         172 Jun  6 16:46 volume-snap-snapshot-ab1a619f-196d-4f58-9a35-2c705a05cacb.img.checksum
-rw-r--r--.   1 root root         191 Jun  8 04:42 volume-snap-snapshot-ab1a619f-196d-4f58-9a35-2c705a05cacb.img.meta
-rw-r--r--.   1 root root 42949672960 Jun  6 15:12 volume-snap-snapshot-d7b5d42f-0111-44a0-b9b7-6bc080a5a809.img
-rw-r--r--.   1 root root         172 Jun  6 16:51 volume-snap-snapshot-d7b5d42f-0111-44a0-b9b7-6bc080a5a809.img.checksum
-rw-r--r--.   1 root root         210 Jun  6 15:12 volume-snap-snapshot-d7b5d42f-0111-44a0-b9b7-6bc080a5a809.img.meta
-rw-r--r--.   1 root root 42949672960 Jun  6 12:43 volume-snap-snapshot-fadec910-b472-45c0-bd0c-d11f0f5b0234.img
-rw-r--r--.   1 root root         172 Jun  6 16:49 volume-snap-snapshot-fadec910-b472-45c0-bd0c-d11f0f5b0234.img.checksum
-rw-r--r--.   1 root root         210 Jun  6 12:43 volume-snap-snapshot-fadec910-b472-45c0-bd0c-d11f0f5b0234.img.meta
-rw-r--r--.   1 root root 42949672960 Jun  7 09:06 volume-snap-snapshot-ffb8c77b-8968-443d-b9e4-d858b9fa5261.img
-rw-r--r--.   1 root root         172 Jun  7 16:03 volume-snap-snapshot-ffb8c77b-8968-443d-b9e4-d858b9fa5261.img.checksum
-rw-r--r--.   1 root root         210 Jun  7 09:06 volume-snap-snapshot-ffb8c77b-8968-443d-b9e4-d858b9fa5261.img.meta
-rw-r--r--.   1 root root         204 Jun  8 04:42 volume.meta
```
grep "" *.meta
```
volume-head-007.img.meta:{"Name":"volume-head-007.img","Parent":"volume-snap-snapshot-23cbf46b-e1f8-41c7-8d21-edbdacdc38a0.img","Removed":false,"UserCreated":false,"Created":"2023-06-08T04:42:09Z","Labels":null}
volume-snap-dailybac-d0c4f62a-8f7a-4522-854e-c754e1dadeb9.img.meta:{"Name":"volume-head-005.img","Parent":"volume-snap-snapshot-71db93c1-d06f-4689-9365-5892a4bfc642.img","Removed":false,"UserCreated":true,"Created":"2023-06-08T04:40:06Z","Labels":{"RecurringJob":"dailybackup"}}
volume-snap-snapshot-0236df7a-8b33-4569-8014-e33d735a4e01.img.meta:{"Name":"volume-head-001.img","Parent":"volume-snap-snapshot-ffb8c77b-8968-443d-b9e4-d858b9fa5261.img","Removed":false,"UserCreated":true,"Created":"2023-06-07T12:03:57Z","Labels":{"RecurringJob":"snapshots"}}
volume-snap-snapshot-23cbf46b-e1f8-41c7-8d21-edbdacdc38a0.img.meta:{"Name":"volume-head-006.img","Parent":"volume-snap-dailybac-d0c4f62a-8f7a-4522-854e-c754e1dadeb9.img","Removed":false,"UserCreated":true,"Created":"2023-06-08T04:42:09Z","Labels":{"RecurringJob":"snapshots"}}
volume-snap-snapshot-488c080c-0b4f-442f-aeec-667cd36f58cb.img.meta:{"Name":"volume-snap-snapshot-488c080c-0b4f-442f-aeec-667cd36f58cb.img","Parent":"volume-snap-snapshot-65bfafe1-9581-496a-81bf-78a3151c658d.img","Removed":false,"UserCreated":true,"Created":"2023-06-06T12:11:14Z","Labels":{"RecurringJob":"snapshots"}}
volume-snap-snapshot-60621c68-3dc8-445d-bc08-f0f3c5587416.img.meta:{"Name":"volume-head-002.img","Parent":"volume-snap-snapshot-0236df7a-8b33-4569-8014-e33d735a4e01.img","Removed":false,"UserCreated":true,"Created":"2023-06-07T15:08:29Z","Labels":{"RecurringJob":"snapshots"}}
volume-snap-snapshot-65bfafe1-9581-496a-81bf-78a3151c658d.img.meta:{"Name":"volume-snap-snapshot-65bfafe1-9581-496a-81bf-78a3151c658d.img","Parent":"volume-snap-snapshot-ab1a619f-196d-4f58-9a35-2c705a05cacb.img","Removed":false,"UserCreated":true,"Created":"2023-06-06T12:04:22Z","Labels":{"RecurringJob":"snapshots"}}
volume-snap-snapshot-71db93c1-d06f-4689-9365-5892a4bfc642.img.meta:{"Name":"volume-snap-snapshot-71db93c1-d06f-4689-9365-5892a4bfc642.img","Parent":"volume-snap-snapshot-60621c68-3dc8-445d-bc08-f0f3c5587416.img","Removed":false,"UserCreated":true,"Created":"2023-06-08T04:35:33Z","Labels":{"RecurringJob":"snapshots"}}
volume-snap-snapshot-ab1a619f-196d-4f58-9a35-2c705a05cacb.img.meta:{"Name":"volume-snap-snapshot-ab1a619f-196d-4f58-9a35-2c705a05cacb.img","Parent":"","Removed":false,"UserCreated":true,"Created":"2023-06-02T15:23:06Z","Labels":{"RecurringJob":"snapshots"}}
volume-snap-snapshot-d7b5d42f-0111-44a0-b9b7-6bc080a5a809.img.meta:{"Name":"volume-head-009.img","Parent":"volume-snap-snapshot-fadec910-b472-45c0-bd0c-d11f0f5b0234.img","Removed":false,"UserCreated":true,"Created":"2023-06-06T15:12:34Z","Labels":{"RecurringJob":"snapshots"}}
volume-snap-snapshot-fadec910-b472-45c0-bd0c-d11f0f5b0234.img.meta:{"Name":"volume-head-008.img","Parent":"volume-snap-snapshot-488c080c-0b4f-442f-aeec-667cd36f58cb.img","Removed":false,"UserCreated":true,"Created":"2023-06-06T12:43:18Z","Labels":{"RecurringJob":"snapshots"}}
volume-snap-snapshot-ffb8c77b-8968-443d-b9e4-d858b9fa5261.img.meta:{"Name":"volume-head-010.img","Parent":"volume-snap-snapshot-d7b5d42f-0111-44a0-b9b7-6bc080a5a809.img","Removed":false,"UserCreated":true,"Created":"2023-06-07T09:06:27Z","Labels":{"RecurringJob":"snapshots"}}
volume.meta:{"Size":42949672960,"Head":"volume-head-007.img","Dirty":true,"Rebuilding":false,"Error":"","Parent":"volume-snap-snapshot-23cbf46b-e1f8-41c7-8d21-edbdacdc38a0.img","SectorSize":512,"BackingFilePath":""}
```

To make it clear:
The nodes are rebooting every morning. The issue only occurs since 1.4.1, not with 1.3.x

**Comment by ejweber** [comment]
Really appreciate your help diagnosing this @pwurbs!

> @innobead Where can I send the support bundle to?

You can send it (or a link) to longhorn-support-bundle@suse.com.

> * The nodes are rebooting every morning. The issue only occurs since 1.4.1, not with 1.3.x

I *think* this is just happenstance, as we have seen it in v1.3.x clusters as well. As you have observed, it is somewhat rare and unpredictable. We believe we have identified the root cause, but it is a race that doesn't often present itself.

> Below the results of the command for one of the affected volumes.

I have reordered the output you provided below to reflect the natural order of the snapshot chain. From what I can tell:

- The replica was 10 GiB when a snapshot was taken at `2023-06-02T15:23:06Z`.
- The replica was 10 GiB when the next snapshot was taken at `2023-06-06T12:04:22Z`.
- The replica was 40 GiB when the next snapshot was taken at `2023-06-06T12:11:14Z`.
- The replica remains 40 GiB.

I'm guessing this volume is configured to be 10 GiB in size. It looks like the node running this replica rebooted around `2023-06-06T12:04:22Z` and a different engine (one responsible for a 40 GiB volume) tried to use it. That engine failed, but not before expanding the replica. While this was a mistake, it did not cause the data on the replica to be incorrect, nor did it cause the replica to fail. The engine, replica, and workload continued to run. Then, at some point the workload scaled down or moved. Especially if the correct replica(s) were down at the time, there is a path for this to lead to only the one, incorrect replica being available in the cluster.

We believe we have the *root cause* fixed in v1.4.2, but there is no automatic recovery path for replicas that were previously broken by an affected version.

- Do you think you were running v1.4.2 at `2023-06-06T12:11:14Z` or v1.4.1?
- Is the expected size of this volume 10 GiB.

**Comment by ejweber** [comment]
There are two (undesirable) fixes here:

1. Scale down the workload consuming the volume, resize it to 40 GiB (in the specific case of this volume), and scale it back up. 
2. Restore from backup (as you have indicated).

**Comment by pwurbs** [comment]
There was no node reboot around 2023-06-06T12:04:22Z. The recurring snapshot job is scheduled for 12:00. 
Rather we daily shut down the nodes in the evening and start them up in the morning. 
And at the first start-up (2023-06-07) after the LH update (2023-06-06) , the issue was visible. 

I started the upgrade from 1.4.1 to 1.4.2 in the affected cluster at 06.06.2023, 13:24 CEST. Then there was chaos in the cluster due to the known issue, exceeding CPU resources while (rolling) upgrading the instance-manager pods. I had to change the resource spec of the instance-manager pods and then revert the value. It took until about 13:45.

Regarding your 1st question: 2023-06-06T12:11:14Z equals 14:11 CEST. So it was after the chaotic LH update process. Thus, Longhorn 1.4.2 ran already. 
I still wonder, why there are 40GB snapshots for the volume?

And yes, the expected size of the affected volume is 10GB. There is another pod with the same name in a different namespace, using a volume of 40GB. This is suspicious, but not the case for the other 2 affected volumes.

Is this a right conclusion?:
Due to any reason there are snapshots with a wrong size. This is firstly only a hidden issue until a (not clear definable) event, e.g. node restart. And only after this event the issue becomes visible, makes the volume stuck and prevents rebuilding replica.

So, shall I investigate for the other affected volumes, if we see the same snapshot situation?
Is there any pattern for which I could search for, to detect volumes with potential issues?

**Comment by pwurbs** [comment]
I checked another affected volume:
- Nominal size 2GB
- "ls -al" in the replica folder shows 10GB size for the replica snapshots. 
- We can't see the change from good to bad size because the snapshot retention is only 2 days.

**Comment by pwurbs** [comment]
The last issue has been solved automatically, maybe due to the next restart. The 30GB replica snapshot vanished.

**Comment by ejweber** [comment]
This is a good (but unfortunate) find. It seems to confirm that there's some other way this issue can occur that we haven't identified yet. Do you have a support bundle from today after all the nodes are up and running (I don't think we've received any support bundles yet)? Unfortunately, with nodes being stopped every evening and started every morning, our support bundles will generally only contain very recent information.

> Is this a right conclusion?:
> Due to any reason there are snapshots with a wrong size. This is firstly only a hidden issue until a (not clear definable) event, e.g. node restart. And only after this event the issue becomes visible, makes the volume stuck and prevents rebuilding replica.

I realized that the issue you linked wasn't the most up-to-date regarding this issue. Please read https://github.com/longhorn/longhorn/issues/5709 if you are inclined. This has typically been caused by a wrong engine incorrectly communicating with a replica, realizing the replica is not large enough, and expanding it. (So a 30GB engine might communicate with a 15GB replica and expand it.) Names and namespaces are generally irrelevant, because it is all happening in the Longhorn backend (in the `longhorn-system` namespace). This does NOT cause the running replica to fail. However, whenever the replica restarts (e.g., during a node reboot), it will cause all sorts of problems. If there are other healthy replicas and the volume is able to get up and running with them (as was probably your case with the self-resolving problem), eventually the broken one fails and a correct one is simply rebuilt.

We have plans to further harden against the possibility of a wrong engine communicating with a replica, but we fixed the only root cause of this we have found so far in v1.4.2.

**Comment by pwurbs** [comment]
This morning, the issue happened again with another volume. Now we have 4 impacted volumes.
Interestingly, there are logs from instance-manager containing the "volume-snap-expand" pattern related to the "new" affected volume, only at yesterday morning, not this morning. 

Anyway, I am going to send you a fresh support bundle today via email. 

Btw, isn't this issue (also the linked issue) not a critical one for any user of Longhorn? Even if there is rare occurrence, the impact could be critical and could lead to huge production impact. You know, risk = probability * impact ;)
So to be honest, I am a bit astonished, that it's classified only as severity/3

**Comment by ejweber** [comment]
@pwurbs, I'll be looking through the support bundle this morning to see what clues it can provide. In the meantime, please do one of the following for your affected volumes.

> There are two (undesirable) fixes here:
> 
> 1. Scale down the workload consuming the volume, resize it to 40 GiB (in the specific case of this volume), and scale it back up.
> 2. Restore from backup (as you have indicated).

In regards to the frequency of this issue, we have only ever managed to recreate it by intentionally hard crashing instance-manager-r pods or intentionally hard rebooting a node while volumes are running. When these unusual events occur, there can be stale stored state that confuses the Longhorn system. My understanding is that you are shutting down your entire cluster each evening. Are you scaling down the workloads consuming Longhorn volumes before doing so? Given what we know, I think it is unlikely for this issue to occur if Longhorn volumes are given the opportunity to detach before nodes are shut down. If you are not allowing this opportunity, I recommend doing so to mitigate the problem while we continue to investigate. If you are allowing all volumes to detach, it is an additional interesting data point.

As far as I know, it's not common for Longhorn users to regularly shut down all of their nodes, so we don't have a guide/documentation for it. We are expanding our testing in this area though. Copying @longhorn/qa for a use case we may want to cover if we don't already.

**Comment by pwurbs** [comment]
@ejweber 
Fix 1. worked for one of the volumes. The other volumes are not so critical, so I save the time right now.

Yes, we are shutting down the nodes every evening. We do it at OS level, giving the container 30sec to terminate. This is done at Docker level, not K8S level. So I don't expect that the volumes are detached before shutting down the node. For us, this is a good test how resilient the cluster is against sudden restarts or other failures. Chaos monkey testing, you know... ;)
But seriously, even if we could detach the nodes before we shutdown the node (what currently is not easily possible), this is the wrong direction, imho. 
Modern systems should be resilient that they always mitigate an unplanned and sudden outage of a single component or even a datacenter. (This is btw part of the definition of "cloud native"). And such outage of single components like a virtual machine will definitely happen, especially in the cloud. You know, there is even no any availability SLO on a virtual machine at the hyperscalers. 
So, even if the probability for a sudden node outage / restart is only little, the impact could be huge (it's about the data). So, I can only try to convince you, that the you should take into account that a node can fail at any time (without detaching the volume ;) ).

**Comment by ejweber** [comment]
@pwurbs, no need to convince me! I completely agree that Longhorn should be engineered to hold up under these circumstances and have every intention of eliminating this issue. In our experience it is very rare (even in the case of an entire cluster shutting down unexpectedly, it may or may not occur in even one volume), so I was interested to understand how you were hitting it so frequently. In addition, as far as we have seen, it is always the result of an instance-manager-r pod going down with replicas still running. 

We will use your use-case as a model to recreate as we move forward. My plan is to finish implementing #5845, which should eliminate all possibility of an engine talking to a replica incorrectly and log any attempts. This should allow me to catch all related errors/races (including whatever one still exists that is allowing you to hit this issue) and backport fixes into older versions as necessary.

> Do I understand it right, that the issue can only be triggered by a more or less sudden restart / shutdown of a node, without waiting for volumes to detach?

Yes. We have not identified any way for the issue to occur except when instance-manager-r pods go down while replicas are still running.

**Comment by pwurbs** [comment]
@ejweber I know... ;)
No additional volumes experienced the issue, strange. Even though we restart the cluster every morning. 
Thx for your clarification and your hard work to improve stability. I think, everything is clear now and we must decide for which cluster we might continue the update ( I still only faces this issue using 1.4.x). Restoring from backup is a proper fix, but only if the volumes are not so big. 
So waiting for your fix and hoping it doesn't take so much time ;)

One another question: We only saw the issue because Longhorn refuses to decrease volumes. But what happens if the case is the other way round? I mean if the instance-manager pod accidentally claims a wrong volume and this leads to an size increase. Will this then automatically executed and we are surprised with ever growing volume sizes?

**Comment by ejweber** [comment]
cc @longhorn/dev 

I was able to trigger this using the modifications I have made for #5845 on master-head. Because of those modifications, the "targeted" replica was not expanded. However, there is clearly still the potential for incorrect expansion in all released versions of Longhorn.

I triggered it in a three node cluster with 50 volumes (150 replicas) all actively mounted to a toy ngnix workload. I used the below script to repeatedly kill an instance-manager pod. It took six iterations before a log message surfaced indicating that the modifications prevented an issue.

```bash
#!/bin/bash

current_time=$(date -u +"%Y-%m-%dT%H:%M:%SZ")

while true; do
    # Execute kubectl -n longhorn-system delete command
    kubectl -n longhorn-system delete --grace-period=1 pod instance-manager-c090b6286e8431380a8cbe71c3fb43ec

    # Wait for 300 seconds
    sleep 300

    # Execute kubectl -n longhorn-system logs and grep commands
    kubectl -n longhorn-system logs -l longhorn.io/component=instance-manager --tail -1 --since-time $current_time | grep -i -e "invalid grpc metadata"
    log_result_1=$?

    kubectl -n longhorn-system logs -l app=longhorn-manager --tail -1 --since-time $current_time | grep -i -e "invalid grpc metadata" -e "incorrect volume name" -e "incorrect instance name"
    log_result_2=$?

    # Check if there was any output from the grep commands
    if [[ $log_result_1 -eq 0 || $log_result_2 -eq 0 ]]; then
        echo "Execution stopped. Trigger keywords found in logs."
        break
    else
        echo "Execution completed successfully."
    fi
done
```

The killed instance-manager pod logged the following:

```
2023-06-27T19:48:22.606088859Z [pvc-059a429b-28f7-4b21-a7cc-45c8632be109-r-cd8811d9] time="2023-06-27T19:48:22Z" level=error msg="Invalid gRPC metadata" clientVolumeName=pvc-70c30a6e-c387-4352-aae1-86f94ad334d2 method=/ptypes.ReplicaService/ReplicaGet serverVolumeName=pvc-059a429b-28f7-4b21-a7cc-45c8632be109
2023-06-27T19:48:22.607799147Z [pvc-c1a6441c-bf6e-4a5d-8a6c-58b02da82938-r-03074ae7] time="2023-06-27T19:48:22Z" level=error msg="Invalid gRPC metadata" clientVolumeName=pvc-6ea5e001-3f33-4f2b-9c03-95271d95dccf method=/ptypes.ReplicaService/ReplicaGet serverVolumeName=pvc-c1a6441c-bf6e-4a5d-8a6c-58b02da82938
```

The longhorn-manager pod on the same node logged the following:

```
2023-06-27T19:48:22.607532292Z time="2023-06-27T19:48:22Z" level=error msg="Failed to rebuild replica 10.42.150.121:10147" controller=longhorn-engine engine=pvc-70c30a6e-c387-4352-aae1-86f94ad334d2-e-799c7bff error="proxyServer=10.42.150.121:8501 destination=10.42.150.121:10065: failed to add replica tcp://10.42.150.121:10147 for volume: rpc error: code = Unknown desc = failed to get replica 10.42.150.121:10147: rpc error: code = FailedPrecondition desc = Incorrect volume name; check replica address" node=eweber-v124-worker-1ae51dbb-4pngn volume=pvc-70c30a6e-c387-4352-aae1-86f94ad334d2
2023-06-27T19:48:22.607555900Z time="2023-06-27T19:48:22Z" level=info msg="Removing failed rebuilding replica 10.42.150.121:10147" controller=longhorn-engine engine=pvc-70c30a6e-c387-4352-aae1-86f94ad334d2-e-799c7bff node=eweber-v124-worker-1ae51dbb-4pngn volume=pvc-70c30a6e-c387-4352-aae1-86f94ad334d2
2023-06-27T19:48:22.607560110Z time="2023-06-27T19:48:22Z" level=info msg="Event(v1.ObjectReference{Kind:\"Engine\", Namespace:\"longhorn-system\", Name:\"pvc-70c30a6e-c387-4352-aae1-86f94ad334d2-e-799c7bff\", UID:\"fb4612d5-6ca8-4417-a829-e65fc1eab29a\", APIVersion:\"longhorn.io/v1beta2\", ResourceVersion:\"52778902\", FieldPath:\"\"}): type: 'Warning' reason: 'FailedRebuilding' Failed rebuilding replica with Address 10.42.150.121:10147: proxyServer=10.42.150.121:8501 destination=10.42.150.121:10065: failed to add replica tcp://10.42.150.121:10147 for volume: rpc error: code = Unknown desc = failed to get replica 10.42.150.121:10147: rpc error: code = FailedPrecondition desc = Incorrect volume name; check replica address"
2023-06-27T19:48:22.608620253Z time="2023-06-27T19:48:22Z" level=error msg="Failed to rebuild replica 10.42.150.121:10137" controller=longhorn-engine engine=pvc-6ea5e001-3f33-4f2b-9c03-95271d95dccf-e-17a35c32 error="proxyServer=10.42.150.121:8501 destination=10.42.150.121:10062: failed to add replica tcp://10.42.150.121:10137 for volume: rpc error: code = Unknown desc = failed to get replica 10.42.150.121:10137: rpc error: code = FailedPrecondition desc = Incorrect volume name; check replica address" node=eweber-v124-worker-1ae51dbb-4pngn volume=pvc-6ea5e001-3f33-4f2b-9c03-95271d95dccf
2023-06-27T19:48:22.608673822Z time="2023-06-27T19:48:22Z" level=info msg="Removing failed rebuilding replica 10.42.150.121:10137" controller=longhorn-engine engine=pvc-6ea5e001-3f33-4f2b-9c03-95271d95dccf-e-17a35c32 node=eweber-v124-worker-1ae51dbb-4pngn volume=pvc-6ea5e001-3f33-4f2b-9c03-95271d95dccf
2023-06-27T19:48:22.608734835Z time="2023-06-27T19:48:22Z" level=info msg="Event(v1.ObjectReference{Kind:\"Engine\", Namespace:\"longhorn-system\", Name:\"pvc-6ea5e001-3f33-4f2b-9c03-95271d95dccf-e-17a35c32\", UID:\"0aacfd87-07ee-47b3-b3bb-bd3a068765c7\", APIVersion:\"longhorn.io/v1beta2\", ResourceVersion:\"52778904\", FieldPath:\"\"}): type: 'Warning' reason: 'FailedRebuilding' Failed rebuilding replica with Address 10.42.150.121:10137: proxyServer=10.42.150.121:8501 destination=10.42.150.121:10062: failed to add replica tcp://10.42.150.121:10137 for volume: rpc error: code = Unknown desc = failed to get replica 10.42.150.121:10137: rpc error: code = FailedPrecondition desc = Incorrect volume name; check replica address"
```

I have a complete support bundle from moments after it occurred, so hopefully I can understand what went wrong.

**Comment by PhanLe1010** [comment]
Great finding @ejweber

**Comment by pwurbs** [comment]
Great! Thx @ejweber 
To make sure, please explain exactly what you mean with "if a volume has already been hit by the inappropriate expansion issue"

**Comment by ejweber** [comment]
Sure @pwurbs. The short answer is: volumes that are already showing as degraded with failing rebuilds and error messages about a parent file size being larger than a child file size, or error messages like `Expansion Error: BUG: The expected size 2147483648 of engine pvc-97135cac-3890-42b1-bc32-80a0edec1b2e-e-6151baeb should not be smaller than the current size 5368709120` will not be fixed. These volumes will need to be fixed with https://github.com/longhorn/longhorn/issues/6078#issuecomment-1591181048.

It is also possible that a single replica for a volume running in your cluster has been expanded by mistake due to this issue. If the following conditions are met, there may be no symptoms yet:

1. It was incorrectly expanded AFTER being added to a running engine (i.e. the volume was already successfully mounted by a workload).
2. Its volume hasn't been unmounted and mounted again since this happened.
3. Another replica hasn't been rebuild from it since this happened.

My hope is that no previously unaffected volumes become affected.

**Comment by pwurbs** [comment]
Yes, I would like to give you feedback, if the update prevents from further expansion errors.
But now I have to wait for Rancher to provide the update in their Helm chart catalog. Unfortunately, this takes time...

**Comment by ejweber** [comment]
> Yes, I would like to give you feedback, if the update prevents from further expansion errors.
But now I have to wait for Rancher to provide the update in their Helm chart catalog. Unfortunately, this takes time...

Sorry about that. Looks like it became available a couple of days ago.

https://github.com/rancher/charts/commit/5700745c4f543cd677711c1ba8d4ada730f11ded

**Comment by pwurbs** [comment]
This morning, the issue happened again, about 2 days after updating to 1.4.3.
![image](https://github.com/longhorn/longhorn/assets/38912613/23ee787b-10a7-46cb-9294-741973696b68)
The volume looked fine so far. But I am not sure if it falls under the above criteria of an already affected volume. Maybe the 14 days old snapshots are affected.
Let's wait some more days until all snapshoots have been created under the new version.

**Comment by ejweber** [comment]
Thanks @pwurbs. If you have time, please capture a support bundle immediately after you notice the issue on a new volume, as well as the `ls -la` and `grep "" *.meta` for each replica directory.

For this volume, do you have multiple healthy replicas? Or just one? If you have multiple healthy replicas, have all of them been expanded?

**Comment by pwurbs** [comment]
I am going to check and provide information for the next occurance

**Comment by PhanLe1010** [comment]
Do we have update @pwurbs ?

**Comment by pwurbs** [comment]
The issue didn't occur again so far. But let's wait...

**Comment by pwurbs** [comment]
It still looks good. For me, the issue has been solved. Thx a lot

**Comment by ejweber** [comment]
That's great news @pwurbs! Closing the issue for now. Please reopen or create a new issue if you observe it again.
